### DISCUSSION TITLE
is it possible to quote field names in a select?

### DISCUSSION THREAD

**Discussion by toddgeist** [odata2ts/odata2ts#336] [Q&A] [answered]
Hello,

A few years back Claris FileMaker added oData access to their server product as another API option to access FileMaker Databases. It is now the most powerful API that FileMaker has. It is very fast, and supports most of oData 4. ( Yes FileMaker still exists and is still in widespread use around the world. )

But FileMaker is a very permissive database.  It is not uncommon to encounter databases in the wild with field names with spaces in the name. For example:

- Invoice ID
- First Name
- created at

Yes, this is not a good idea, but there are. thousands of these databases out there that have been in use for a long time so we encounter them all the time. 

When using odata2ts, If I allow renaming, the field names are converted to the naming strategy I set up, which is fine. But if I try to use one of these fields in a Select, it doesn't work because the field/property names are not quoted in the Select in the URL.  For example, this output would work

`select="First Name","Invoice ID"`

But the query builder creates 

`select=First Name,Invoice Id`

Which does not work.

Is there a way to configure odata2ts to quote those field name properties?

I know this is asking to support crumby field naming. But I believe the oData spec does allow for this and that quoting is the correct thing to do. ( I could be wrong about this ).

Thanks for your help

Todd

**Comment by texttechne** [comment]
Hi @toddgeist,
spaces within property names are actually non-conformant, if I get the spec right.
See [this StackOverflow issue](https://stackoverflow.com/questions/19007765/querying-fields-with-name-containing-whitespace-using-odata) including the XML definition.

Nevertheless, we can build in a workaround. But first we need to make sure, which way to go. 

So:
* In the SO issue, they used `_x0020_` which looks to me to be quite Sharepoint specific. Or does this work in your situation as well?
* Looks awkward to use double quotes to be honest. Can you find something in the spec?

**Reply by toddgeist** [reply]
I tried _x0020_ instead of spaces, but that doesn't work.  The only way  that does work is double quotes. Like this

`?select="external id", status, "id"`

**Comment by toddgeist** [comment]
Thanks for the response and the suggestions.  I will investigate further and get back you to you.

I am beginning to think that this implementation may have other other weird things that may get in the way. For example, for batch updates, we need to use multi-part mixed, not JSON as the content type.   Yuck!

Thanks for your help. This library is very powerful. I hope we can figure out how to use it for our case.

**Comment by texttechne** [comment]
> I am beginning to think that this implementation may have other other weird things that may get in the way. For example, for batch updates, we need to use multi-part mixed, not JSON as the content type. Yuck!

See [Spec](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_BatchRequests). Looks like it's working as it is supposed to be. Try sending the `Accept` header with `application/json` or set the `$format` query parameter accordingly.

However, batch processing isn't implemented in odata2ts. One of my colleagues toyed around with `odata-batch-request`. odata2ts only served to get the correct URLs in this scenario...

**Reply by toddgeist** [reply]
sorry this wasn't clear.  It's the oData service implementation that has other issues, not your library. This odata service uses multipart/mixed for the Batch format.

**Comment by texttechne** [comment]
Hi @toddgeist,

converted the issue into a discussion. Feels more appropriate...

**Comment by texttechne** [comment, marked as answer]
Hi @toddgeist,
you can now `select` and `expand` custom/unknown properties. 
So now you are able to specify what you want: see #356
